### PR TITLE
Optimize string replacements in camel/kebab case transformations

### PR DESCRIPTION
**Pull request by waiter-melon**
Pre-compiling regular expressions drastically improves camel-case and kebab-case transformations.

Here a benchmark focused on `snakeCase` function. 
Before:
``
[info] KeyCasingBenchmark.snakeCase  thrpt   10    829556.926 ± 59956.575  ops/s
``
After:
``
[info] KeyCasingBenchmark.snakeCase  thrpt   10  1861369.701 ± 52228.706  ops/s
``

### CODE REVIEW

File: modules/generic-extras/src/main/scala/io/circe/generic/extras/Configuration.scala [hunk 1]
```diff
@@ -40,17 +42,21 @@ final case class Configuration(
 }
 
 final object Configuration {
+
   val default: Configuration = Configuration(Predef.identity, Predef.identity, false, None)
+  val basePattern: Pattern = Pattern.compile("([A-Z]+)([A-Z][a-z])")
```
Comment (travisbrown) [inline modules/generic-extras/src/main/scala/io/circe/generic/extras/Configuration.scala]: Could you make these private? They shouldn't be needed outside of the implementations.

Comment (plokhotnyuk) [reply modules/generic-extras/src/main/scala/io/circe/generic/extras/Configuration.scala]: Here are fast implementations of enforcing transformations for snake-case, kebab-case, and camelCase: https://github.com/plokhotnyuk/jsoniter-scala/blob/master/jsoniter-scala-macros/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/macros/JsonCodecMaker.scala#L52-L98

Please, fill free to adopt them for your requirements.

**Comment by waiter-melon** [comment]
@plokhotnyuk fast implementations are indeed much faster:

`KeyCasingBenchmark.snakeCase  thrpt  10457001.059  ops/s`

The thing is that they produce slightly different outputs, and I'm not sure it fits your needs:

```scala
// plokhotnyuk implementation
Configuration.snakeCaseTransformation("SnakeCase")
Configuration.snakeCaseTransformation("snakeCase")
Configuration.snakeCaseTransformation("snakecase")
Configuration.snakeCaseTransformation("snake_case")
Configuration.snakeCaseTransformation("snake-case")
Configuration.snakeCaseTransformation("SNAKECASE")
Configuration.snakeCaseTransformation("")
Configuration.snakeCaseTransformation("sNaKeCaSe")
Configuration.snakeCaseTransformation("SNAKEcase")
Configuration.snakeCaseTransformation("SnakecasE")

//res0: String = _snake_case
//res1: String = snake_case
//res2: String = snakecase
//res3: String = snake_case
//res4: String = snake_case
//res5: String = snakecase
//res6: String =
//res7: String = s_na_ke_ca_se
//res8: String = snak_ecase
//res9: String = _snakecas_e

javaRegexSnakeCaseTransformation("SnakeCase")
javaRegexSnakeCaseTransformation("snakeCase")
javaRegexSnakeCaseTransformation("snakecase")
javaRegexSnakeCaseTransformation("snake_case")
javaRegexSnakeCaseTransformation("snake-case")
javaRegexSnakeCaseTransformation("SNAKECASE")
javaRegexSnakeCaseTransformation("")
javaRegexSnakeCaseTransformation("sNaKeCaSe")
javaRegexSnakeCaseTransformation("SNAKEcase")
javaRegexSnakeCaseTransformation("SnakecasE")

//res10: String = snake_case
//res11: String = snake_case
//res12: String = snakecase
//res13: String = snake_case
//res14: String = snake-case
//res15: String = snakecase
//res16: String =
//res17: String = s_na_ke_ca_se
//res18: String = snak_ecase
//res19: String = snakecas_e

// plokhotnyuk implementation
Configuration.kebabCaseTransformation("KnakeCase")
Configuration.kebabCaseTransformation("kebabCase")
Configuration.kebabCaseTransformation("kebabcase")
Configuration.kebabCaseTransformation("kebab_case")
Configuration.kebabCaseTransformation("kebab-case")
Configuration.kebabCaseTransformation("KEBABCASE")
Configuration.kebabCaseTransformation("")
Configuration.kebabCaseTransformation("kEbAbCaSe")
Configuration.kebabCaseTransformation("KEBABcase")
Configuration.kebabCaseTransformation("KebabcasE")

res20: String = -knake-case
//res21: String = kebab-case
//res22: String = kebabcase
//res23: String = kebab-case
//res24: String = kebab-case
//res25: String = kebabcase
//res26: String =
//res27: String = k-eb-ab-ca-se
//res28: String = keba-bcase
//res29: String = -kebabcas-e

javaRegexKebabCaseTransformation("KnakeCase")
javaRegexKebabCaseTransformation("kebabCase")
javaRegexKebabCaseTransformation("kebabcase")
javaRegexKebabCaseTransformation("kebab_case")
javaRegexKebabCaseTransformation("kebab-case")
javaRegexKebabCaseTransformation("KEBABCASE")
javaRegexKebabCaseTransformation("")
javaRegexKebabCaseTransformation("kEbAbCaSe")
javaRegexKebabCaseTransformation("KEBABcase")
javaRegexKebabCaseTransformation("KebabcasE")

//res30: String = knake-case
//res31: String = kebab-case
//res32: String = kebabcase
//res33: String = kebab_case
//res34: String = kebab-case
//res35: String = kebabcase
//res36: String =
//res37: String = k-eb-ab-ca-se
//res38: String = keba-bcase
//res39: String = kebabcas-e
```

Let me know how you want me to proceed?

**Comment by plokhotnyuk** [comment]
@waiter-melon thank you for your feedback! leading `_` looks like a bug - I will fix it

**Comment by waiter-melon** [comment]
Could we discard faster implementation and just pre-compilation of regular expressions? We should include this in the next release I guess.

**Comment by travisbrown** [comment]
Thanks, @waiter-melon—these commits will make it into 0.11.0 via #940.

**Pull request closed without merging**